continuous-unregular-load.py

```diff
@@ -0,0 +1,278 @@
+
+import random
+
+import requests
+
+log = logging.getLogger()
+logging.basicConfig(
+    level=logging.INFO,
+    format="%(asctime)s.%(msecs)03d %(levelname)s %(threadName)s: %(message)s",
+    datefmt="%y%m%d-%H:%M:%S",
+)
+
+RANDOM_RUN_IDS = [uuid.uuid4().hex for _ in range(10)]
+
+
+# base_url = "http://127.0.0.1:8000/api"
+# if os.environ.get("CONBENCH_BASE_URL"):
+base_url = f"{os.environ['CONBENCH_BASE_URL']}"
+
+
+session = requests.Session()
+
+# The number is the desired target rate, in 1/s.
+
+targets = [
+    # ("/api/ping/", "GET", 0.5),
+    # ("/", "GET", 1.0),
+    # (f"/runs/{RANDOM_RUN_IDS[0]}/", "GET", 0.1),
+    # (f"/api/compare/runs/{RANDOM_RUN_IDS[0]}...{RANDOM_RUN_IDS[3]}/", "GET", 1.0),
+    ("/api/benchmark-results/", "POST", 10.0),
+    # (
+    #     "/compare/benchmarks/fd773f3414b04ca783c97925fe05c2c5...40872ceb41094f1ea509b19ccb762fb8/",
+    #     "GET",
+    #     0.2,
+    # ),
+    # ("/batches/0d32f00ea4884aaea10454cf61f7e91f/", "GET", 0.1),
+    # ("/benchmarks/4a175cbb62cb473389e78174e5dc9991/", "GET", 0.1),
+    # ("/benchmarks/35b0b18277a24bbda262ecd66bf5ad42/", "GET", 0.5),
+    # ("/login/", "GET", 0.05),
+]
+
+
+process_launch_time = time.monotonic()
+last_time_i_did_that = {}
+
+# Large maxsize just to pre-pan for the fact that there might be a higher work
+# creation rate than a fixed-size executor pool can process, and this case
+# needs to be handled by code (back off, instead of having infinite memory
+# usage growth).
+q = queue.Queue(maxsize=1000)
+
+
+INIT_SHUTDOWN = False
+
+HTTP_REQUESTS_MADE_COUNTER = 0
+
+
+def main():
+    global INIT_SHUTDOWN
+
+    register()
+    login()
+
+    # Start the task creator thread (infinite loop)
+    t_creator = Thread(target=task_creator)
+    t_creator.start()
+
+    # Start N task consumer threads, each also running an infinite loop.
+    t_consumers = [Thread(target=task_consumer) for _ in range(10)]
+
+    for thread in t_consumers:
+        thread.start()
+
+    # main thread, hang out a bit.
+    while True:
+        log.debug("main thread: bored.. hello")
+
+        if (HTTP_REQUESTS_MADE_COUNTER % 50) == 0:
+            log.info("HTTP requests made so far: %s", HTTP_REQUESTS_MADE_COUNTER)
+
+        try:
+            time.sleep(0.2)
+        except KeyboardInterrupt:
+            log.info("caught SIGINT, initiate shutdown")
+            INIT_SHUTDOWN = True
+            break
+
+    log.info("joining creator thread")
+    t_creator.join()
+
+    for thread in t_consumers:
+        log.info("joining consumer thread %s", thread)
+        thread.join()
+
+
+def task_consumer():
+    while True:
+        if INIT_SHUTDOWN:
+            log.info("consumer thread: exit")
+            return
+        _task_consumer_iteration()
+        time.sleep(0.001)
+
+
+def task_creator():
+    while True:
+        if INIT_SHUTDOWN:
+            log.info("creator thread: exit")
+            return
+
+        _task_creator_iteration()
+        time.sleep(0.01)
```

Comment:
Oh, I was recently looking for this module here locally on my machine and didn't find it anymore. Happy that I did push it to this branch. I think it's a useful load generation helper that I once wrote here.